File: 1971-find-if-path-exists-in-graph/1971-find-if-path-exists-in-graph.py
class Solution:
    def validPath(self, n: int, edges: List[List[int]], source: int, destination: int) -> bool:
        def find(x):
            if x != p[x]:
                p[x] = find(p[x])
            return p[x]
    
        def union(a, b):
            ap = find(a)
            bp = find(b)
            p[ap] = bp
            
        p = [i for i in range(n)]
        for u, v in edges:
            if find(u) != find(v):
                union(u, v)
                
        return find(source) == find(destination)

File: 1971-find-if-path-exists-in-graph/1971-find-if-path-exists-in-graph.rs
use std::collections::VecDeque;

impl Solution {
    pub fn valid_path(n: i32, edges: Vec<Vec<i32>>, source: i32, destination: i32) -> bool {
        let n = n as usize;
        let source = source as usize;
        let mut graph: Vec<Vec<usize>> = vec![vec![]; n];
        
        for edge in edges {
            let u = edge[0] as usize;
            let v = edge[1] as usize;
            
            graph[u].push(v);
            graph[v].push(u);
        }
        
        let mut q: VecDeque<usize> = VecDeque::from([source as usize]);
        let mut visited: Vec<bool> = (0..n).map(|x| {
            if x == source {
                true
            } else {
                false
            }
        }).collect();
        
        while let Some(now) = q.pop_front() {
            if now == destination as usize {
                return true;
            }
            
            for next in graph[now].iter() {
                if visited[*next] {
                    continue;
                }
                
                visited[*next] = true;
                q.push_back(*next);
            }
        }
        false
    }
}